cmd/fleet/main.go

```diff
@@ -161,141 +83,160 @@ func NewCommand(version string) *cobra.Command {
 	return cmd
 }
 
-type fleetServer struct {
-	cfg *config.Config
-
-	es *es.Client
-	sv saved.CRUD
-
-	lock           sync.Mutex
-	profilerCancel context.CancelFunc
-	serverCancel   context.CancelFunc
-	errCh          chan error
+type FleetServer struct {
+	cfg   *config.Config
+	cfgCh chan *config.Config
 }
 
 // NewFleetServer creates the actual fleet server service.
-func NewFleetServer(ctx context.Context, cfg *config.Config) (*fleetServer, error) {
-	es, err := es.New(ctx, cfg)
-	if err != nil {
-		return nil, err
-	}
-	sv := saved.NewMgr(es.Bulk(), savedObjectKey())
-	return &fleetServer{
-		cfg: cfg,
-		es:  es,
-		sv:  sv,
+func NewFleetServer(cfg *config.Config) (*FleetServer, error) {
+	return &FleetServer{
+		cfg:   cfg,
+		cfgCh: make(chan *config.Config, 1),
 	}, nil
 }
 
 // Run runs the fleet server.
-func (f *fleetServer) Run(ctx context.Context) error {
-	// Initial indices bootstrapping, needed for agents actions development
-	// TODO: remove this after the indices bootstrapping logic implemented in ES plugin
-	err := esboot.EnsureESIndices(ctx, f.es)
-	if err != nil {
-		return err
+func (f *FleetServer) Run(ctx context.Context) error {
+	var curCfg *config.Config
+	newCfg := f.cfg
+
+	// To avoid repeating the same code for policy server and http server
+	// The original code required them to be restarted independently,
+	// to avoid restarting HTTP server when the profiler is enabled/disabled
+	start := func(cn context.CancelFunc,
+		wg *sync.WaitGroup,
+		runfn runner.RunFunc,
+		errCh chan<- error,
+	) context.CancelFunc {
+		if cn != nil {
+			cn()
+		}
+		wg.Wait()
+		cx, cn := context.WithCancel(ctx)
+		runner.Start(cx, wg, runfn, func(er error) {
+			if er != nil {
+				errCh <- er
+			}
+		})
+		return cn
 	}
 
-	// profiler restarts on reload
-	profilerCtx, profilerCancel := context.WithCancel(ctx)
-	err = profile.RunProfiler(profilerCtx, f.cfg.Inputs[0].Server.Profile.Bind)
-	if err != nil {
-		profilerCancel()
-		return err
+	var (
+		proCancel, srvCancel context.CancelFunc
+		proWg, srvWg         sync.WaitGroup
+	)
+
+	for {
+		ech := make(chan error, 2)
+
+		if curCfg == nil || curCfg.Inputs[0].Server.Profile.Bind != newCfg.Inputs[0].Server.Profile.Bind {
```

Comment:
When will `curCfg` be `nil`? Seems that it should never be nil.

Reply:
it's on initial pass when the server is started for the first time.
the newCfg is assigned to f.cfg initially (line 102) and then we get the updates from the Reload channel (line 149)
otherwise following the same logic comparing the config params as it was before in the Reload func,
but doing start/restart in the scope of the func.

Reply:
Still confused on when it will be `nil` then. Look at it, if `curCfg = f.cfg` on initial `Run` then its always set.

Reply:
```
newCfg := f.cfg
```
curCfg in my interpretation here is the "currently applied/running config" (only after it applied)

maybe can reverse the naming if confusing, let me know

Reply:
Ah, I see. I was mixing up `curCfg` with `newCfg`. Looks good.